pegen/pegen.c

```diff
@@ -1677,70 +1613,1156 @@ parsestr(Parser *p, const char *s, int *bytesmode, int *rawmode, PyObject **resu
-        int this_bytesmode = PyBytes_CheckExact(s);
+        Token *t = asdl_seq_GET(strings, i);
+
+        int this_bytesmode;
+        int this_rawmode;
+        PyObject *s;
+        const char *fstr;
+        Py_ssize_t fstrlen = -1;
 
+        char *this_str = PyBytes_AsString(t->bytes);
+
+        if (parsestr(p, this_str, &this_bytesmode, &this_rawmode, &s, &fstr, &fstrlen) != 0) {
+            return NULL;
+        }
+
+        /* Check that we are not mixing bytes with unicode. */
         if (i != 0 && bytesmode != this_bytesmode) {
             raise_syntax_error(p, "cannot mix bytes and nonbytes literals");
+            Py_XDECREF(s);
             goto error;
         }
-
         bytesmode = this_bytesmode;
 
-        if (bytesmode) {
-            PyBytes_Concat(&final_str, s);
-            if (!final_str) {
+        if (fstr != NULL) {
+            assert(s == NULL && !bytesmode);
+
+            int result = FstringParser_ConcatFstring(p, &state, &fstr, fstr+fstrlen,
+                                                     this_rawmode, 0, t, last);
+            if (result < 0) {
                 goto error;
             }
-        }
-        else {
-            kind_unicode |= (cons->v.Constant.kind != NULL);
-            PyUnicode_Append(&final_str, s);
-            if (!final_str) {
-                goto error;
+        } else {
+            /* String or byte string. */
+            assert(s != NULL && fstr == NULL);
+            assert(bytesmode ? PyBytes_CheckExact(s) : PyUnicode_CheckExact(s));
+
+            if (bytesmode) {
+                if (i == 0) {
+                    bytes_str = s;
+                } else {
+                    PyBytes_ConcatAndDel(&bytes_str, s);
+                    if (!bytes_str) {
+                        goto error;
+                    }
+                }
+            } else {
+                /* This is a regular string. Concatenate it. */
+                if (FstringParser_ConcatAndDel(&state, s) < 0) {
+                    goto error;
+                }
             }
         }
     }
 
-    if (kind_unicode) {
-        // TODO: Intern this string when we decide how we will
-        // handle static constants in the module.
-        u_kind = new_identifier(p, "u");
-    }
-    if (PyArena_AddPyObject(p->arena, final_str) < 0) {
-        Py_DECREF(final_str);
-        return NULL;
+    if (bytesmode) {
+        if (PyArena_AddPyObject(p->arena, bytes_str) < 0) {
+            goto error;
+        }
+        return Constant(bytes_str, NULL, first->lineno, first->col_offset,
+                        last->end_lineno, last->end_col_offset, p->arena);
     }
-    return _Py_Constant(final_str, u_kind, EXTRA_EXPR(first, last));
+
+    return FstringParser_Finish(p, &state, first, last);
 
 error:
-    Py_XDECREF(final_str);
+    Py_XDECREF(bytes_str);
+    FstringParser_Dealloc(&state);
     return NULL;
 }
+
+void *
+string_token(Parser *p)
+{
+    return expect_token(p, STRING);
+}
```

Comment:
Seriously, this file is now over 2700 lines long? I worry that by the time we're done, it'll be as long as ast.c (nearly 6000). That would make arguing for this harder -- I had hoped to eliminate most of ast.c, and it looks like I vastly overestimated the possible win there... :-(

And is the f-string code in ast.c really over 1000 lines? At the very least it's probably a good idea to move it to its own file (even if that means we'll have some more public symbols).

Reply:
That's a bit sad that f-strings made things so much uglier. 1700 lines wasn't that much, but now it really is more chaotic than before. At least, I think the code is somewhat cleaner.

> At the very least it's probably a good idea to move it to its own file (even if that means we'll have some more public symbols).

Agreed!